### DISCUSSION TITLE
use boost special functions in Function object definition

### DISCUSSION THREAD

**Discussion by chunhuizhao478** [idaholab/moose#26172] [Q&A General]
Hi all, I would like to use `expint` function in `boost/math/special_functions/expint.hpp`, after include the header file in the Function object:

`#include <boost/math/special_functions/expint.hpp>`

It throws out the file-not-found error:

```
/Users/andyz/projects/farms/src/functions/cdbm/InitialStressXYPressure.C:4:10: fatal error: 'boost/math/special_functions/expint.hpp' file not found
#include <boost/math/special_functions/expint.hpp>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

As I searched in the MOOSE source code, there are cases including boost special functions, so I wonder what I missed here, thanks!

**Comment by lindsayad** [comment]
If you look at the uses of boost special functions in MOOSE, they are guarded with
```
#ifdef LIBMESH_HAVE_EXTERNAL_BOOST
```
libMesh does have some boost headers. However, it does not have them all. In particular it does not have `expint.hpp`. So you would need to configure libMesh pointing to an external boost that has that header in order to be able to use it.

@roystgnr why do we have an "incomplete" boost?

**Reply by chunhuizhao478** [reply]
So I wonder if there is still a way to directly add `expint` function in MOOSE without too much complexity, what I currently do is to compute the expression contains `expint` ouside MOOSE using Python and read the time series data file into MOOSE using `PiecewiseMultilinear`.

**Reply by GiudGiud** [reply]
if it s just for your app / branch you can add a contrib then modify the Makefile

**Reply by chunhuizhao478** [reply]
Hi @GiudGiud, could you elaborate more? I'm not familiar with the modifications I should do, thanks!

**Reply by roystgnr** [reply]
Complete boost is so huge (I just downloaded a tarball that expands to 900MB, 50x the subset we have) that boost itself includes `bcp`, a tool for creating incomplete subsets of boost.

Originally the libMesh logs claim Ben grabbed a subset of boost because we wanted `shared_ptr` and C++11 wasn't out yet.  Later we started optionally using their pool allocator ... and I'm not sure what else.  The ability to configure `--with-boost=/some/external/boost` is the recommended way to use arbitrary parts of boost.

**Reply by GiudGiud** [reply]
ah me neither. 
But you should be able to place a header from an external library into a contrib folder in your app, then in the Makefile add some logic so that it gets include in the include list
